File: lib/auth/register_view_model.dart
import 'package:flutter/foundation.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:diplomovka/auth/auth_service.dart';
import 'package:diplomovka/auth/register_user_use_case.dart';

class RegisterViewModel extends ChangeNotifier {
  final RegisterUserUseCase _registerUserUseCase;

  RegisterViewModel(AuthService authService)
      : _registerUserUseCase = RegisterUserUseCase(authService);

  String? _error;
  bool _isLoading = false;

  String? get error => _error;
  bool get isLoading => _isLoading;

  Future<void> register(String email, String password) async {
    if (_isLoading) return;

    _isLoading = true;
    _error = null;
    notifyListeners();

    try {
      await _registerUserUseCase.execute(email, password);

    } on FirebaseAuthException catch (e) {
      if (e.code == 'weak-password') {
        _error = 'Heslo je príliš slabé.';
      } else if (e.code == 'email-already-in-use') {
        _error = 'Účet s týmto e-mailom už existuje.';
      } else if (e.code == 'invalid-email') {
        _error = 'Neplatný formát e-mailu.';
      } else {
        _error = 'Chyba pri registrácii: ${e.message}';
      }
    } catch (e) {
      _error = 'Nastala neočakávaná chyba. Skúste to znova.';
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  void clearError() {
    _error = null;
    notifyListeners();
  }
}

File: app/src/main/java/com/example/diplomovka_kotlin/viewmodel/RegisterViewModel.kt
package com.example.diplomovka_kotlin.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.example.diplomovka_kotlin.data.services.AuthService
import com.example.diplomovka_kotlin.domain.RegisterUserUseCase
import com.google.firebase.auth.FirebaseAuthException

class RegisterViewModel(application: Application) : AndroidViewModel(application) {

    private val authService = AuthService(application)
    private val registerUseCase = RegisterUserUseCase(authService)

    private val _isLoading = MutableLiveData(false)
    val isLoading: LiveData<Boolean> = _isLoading

    private val _error = MutableLiveData<String?>(null)
    val error: LiveData<String?> = _error

    suspend fun register(email: String, password: String): Boolean {
        _isLoading.value = true
        return try {
            registerUseCase.execute(email, password)
            true
        } catch (e: FirebaseAuthException) {
            _error.value = when (e.errorCode) {
                "email-already-in-use" -> "Tento e-mail je už zaregistrovaný."
                "invalid-email"        -> "Neplatný formát e-mailu."
                "weak-password"        -> "Heslo je príliš slabé."
                else -> "Chyba pri registrácii: ${e.message ?: e.errorCode}"
            }
            false
        } finally {
            _isLoading.value = false
        }
    }

    fun clearError() { _error.value = null }
}
